runtime/jvmti/jvmtiHelpers.c

```diff
@@ -789,69 +789,71 @@ getVirtualThreadState(J9VMThread *currentThread, jthread thread)
 	Assert_JVMTI_notNull(thread);
 	Assert_JVMTI_mustHaveVMAccess(currentThread);
 	if (JVMTI_ERROR_NONE == getVMThread(currentThread, thread, &targetThread, FALSE, FALSE)) {
-		j9object_t vThreadObject = J9_JNI_UNWRAP_REFERENCE(thread);
-		jint vThreadState = (jint) J9VMJAVALANGVIRTUALTHREAD_STATE(currentThread, vThreadObject);
-		/* The mapping from JVMTI_VTHREAD_STATE_XXX to JVMTI_JAVA_LANG_THREAD_STATE_XXX is based on j.l.VirtualThread.threadState(). */
-		switch (vThreadState) {
-		case JVMTI_VTHREAD_STATE_NEW:
-			rc = JVMTI_JAVA_LANG_THREAD_STATE_NEW;
-			break;
-		case JVMTI_VTHREAD_STATE_STARTED:
-		{
-			JNIEnv *env = (JNIEnv *)currentThread;
-			jfieldID fid = NULL;
-			jclass jlThread = NULL;
-
-			vm->internalVMFunctions->internalExitVMToJNI(currentThread);
-			jlThread = (*env)->FindClass(env, "java/lang/Thread");
-			if (NULL != jlThread) {
-				fid = (*env)->GetFieldID(env, jlThread, "container", "Ljdk/internal/vm/ThreadContainer;");
-			}
-			if ((NULL != fid)
-				&& (NULL == (*env)->GetObjectField(env, thread, fid))
-			) {
+		if (NULL != targetThread) {
+			vm->internalVMFunctions->haltThreadForInspection(currentThread, targetThread);
+			rc = getThreadState(currentThread, targetThread->carrierThreadObject);
+			vm->internalVMFunctions->resumeThreadForInspection(currentThread, targetThread);
+		} else {
+			j9object_t vThreadObject = J9_JNI_UNWRAP_REFERENCE(thread);
+			jint vThreadState = (jint) J9VMJAVALANGVIRTUALTHREAD_STATE(currentThread, vThreadObject);
```

Comment:
Other places don't assume `NULL == targetThread` signals a virtual thread: why is that assumption safe here?

Reply:
getVirtualThreadState is only called from one location, which verifies that the input thread parameter is a virtual thread instance.
https://github.com/eclipse-openj9/openj9/blob/f06248f89427d0509360ec6a1df74e0c06436072/runtime/jvmti/jvmtiThread.c#L71-L76